### PR TITLE
Introduce `digdag projects` cmd

### PR DESCRIPTION
**Pull request by hiroyuki-sato**
This PR implement `digdag projects` command. It outputs the project list.
Current API has '/api/projects'. But CLI doesn't exist.

@muga, @yoyama What do you think?

If this proposes command useful, I'll add a test.
Does it integration test?

### CODE REVIEW

**Comment by yoyama** [comment]
@hiroyuki-sato Thanks for your proposal.
Good idea! please continue it.
Just my idea, display not only name, but also another data suche as attempts command may be helpful for users.

**Comment by sonots** [comment]
I want all info which /projects API returns such as created_at, updated_at.
Either of with or without an option is fine.

Also, I thought the API returns deleted_at. 
So, adding an option to show deleted projects or not would be useful.

**Comment by hiroyuki-sato** [comment]
@yoyama, @sonots 
Thank you for the comment.

What do you think about this output?
And do you know whether '/api/projects' returns deleted projects?
It seems that that API doesn't return it.
I couldn't get deleted project lists.

```
./pkg/digdag-0.9.36-SNAPSHOT.jar projects
2019-03-26 17:42:07 +0900: Digdag v0.9.36-SNAPSHOT
fuga
  id: 2
  revision: 9181e423-394c-445e-8434-141dd5671638
  archive type: db
  project created at: 2019-03-26T08:40:10Z
  revision updated at: 2019-03-26T08:40:10Z
hoge
  id: 3
  revision: db2a2658-cf43-4076-9bd1-b134ace63c5b
  archive type: db
  project created at: 2019-03-26T08:42:03Z
  revision updated at: 2019-03-26T08:42:03Z

Use `digdag workflows <project-name>` to show details.
```

**Comment by yoyama** [comment]
@hiroyuki-sato It looks like good.
If API does not support deleted projects, I guess you had better separate it from this PR.

**Comment by hiroyuki-sato** [comment]
@yoyama I'm not sure whether API gets deleted project or not.
But It seems that the API filter deleted the project
https://github.com/treasure-data/digdag/blob/master/digdag-server/src/main/java/io/digdag/server/rs/ProjectResource.java#L275

I removed these codes.

**Comment by sonots** [comment]
Outputs would be better to be like “digdag sessions” or “digdag schedules”.

**Comment by hiroyuki-sato** [comment]
@yoyama, @sonots 

Current implementation referenced `digdag push` command.
What do you think this sample output?


```
Projects:
  project: foo
  id: 1
  revision: 03d00489-501d-4ca7-a76d-70c900549f91
  archive type: db
  project created at: 2019-03-27 10:57:56 +0900
  revision updated at: 2019-03-27 10:57:59 +0900

  project: bar
  id: 2
  revision: 03d00489-501d-4ca7-a76d-70c900549f91
  archive type: db
  project created at: 2019-03-27 10:58:56 +0900
  revision updated at: 2019-03-27 10:59:59 +0900

Use `digdag workflows <project-name>` to show details.
```



## sample output 

`digdag sessions `

```
Sessions:
  session id: 1
  attempt id: 1
  uuid: 34bf1800-ed93-440e-bddf-02439d9a92c3
  project: foo
  workflow: aa
  session time: 2019-03-27 01:57:56 +0000
  retry attempt name:
  params: {}
  created at: 2019-03-27 10:57:56 +0900
  kill requested: false
  status: success

  session id: 2
  attempt id: 2
  uuid: bb6695e9-28a0-479d-9e18-fe5fe4a6004b
  project: bar
  workflow: aa
  session time: 2019-03-27 01:58:02 +0000
  retry attempt name:
  params: {}
  created at: 2019-03-27 10:58:03 +0900
  kill requested: false
  status: running

```

**Comment by sonots** [comment]
lgtm

**Comment by hiroyuki-sato** [comment]
@yoyama, @sonots 

Fixed. Please take a look when you have time.

```
Projects
  name: hoge
  id: 1
  revision: b2451942-59fa-43ab-9f71-5929716ace02
  archive type: db
  project created at: 2019-03-27 17:20:11 +0900
  revision updated at: 2019-03-27 17:20:11 +0900

  name: fuga
  id: 2
  revision: a047baca-83d3-4d7f-8b10-c39abc7b1b45
  archive type: db
  project created at: 2019-03-27 17:20:17 +0900
  revision updated at: 2019-03-27 17:20:17 +0900

Use `digdag workflows <project-name>` to show details.
```

**Comment by yoyama** [comment]
@hiroyuki-sato Thanks, GJ! 
Could you add test? After you finish work, I would like to review.

**Comment by hiroyuki-sato** [comment]
@yoyama, Thank you for the comment.
I added simple Integration test. Please take a look when you have time.

File: digdag-tests/src/test/java/acceptance/CliShowIT.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package acceptance;
+
+import io.digdag.client.DigdagClient;
```
Comment (yoyama) [inline digdag-tests/src/test/java/acceptance/CliShowIT.java]: It looks like not used.

File: digdag-tests/src/test/java/acceptance/CliShowIT.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        }
+
+        CommandStatus showProjectStatus = main("projects",
+                "-c", config.toString(),
+                "-e", server.endpoint());
+
+        assertThat(showProjectStatus.errUtf8(), showProjectStatus.code(), is(0));
+        assertThat(showProjectStatus.outUtf8(), containsString(" name: foo"));
+        assertThat(showProjectStatus.outUtf8(), containsString(" name: bar"));
+        ;
```
Comment (yoyama) [inline digdag-tests/src/test/java/acceptance/CliShowIT.java]: Needless semicolon?

**Comment by yoyama** [comment]
@hiroyuki-sato Thank you for your contribution.
I wrote a few comments. Please check them.
In addition, could you support  "projects <project name>" if possible?
Other commands 'workflows', 'sessions' support argument of name to show one data only.
So I guess supporting it in 'projects' command is also better.

File: digdag-tests/src/test/java/acceptance/CliShowIT.java (hunk 1 above)
Comment (hiroyuki-sato) [reply digdag-tests/src/test/java/acceptance/CliShowIT.java]: Thanks!, Removed

File: digdag-tests/src/test/java/acceptance/CliShowIT.java (hunk 2 above)
Comment (hiroyuki-sato) [reply digdag-tests/src/test/java/acceptance/CliShowIT.java]: Thanks!, Removed

**Comment by hiroyuki-sato** [comment]
@yoyama Thank you for the comment. I fixed your comments.

About projects support, does this mean like the following?

`digdag projects`

```
Projects
  name: hoge
  id: 1
  revision: b2451942-59fa-43ab-9f71-5929716ace02
  archive type: db
  project created at: 2019-03-27 17:20:11 +0900
  revision updated at: 2019-03-27 17:20:11 +0900

  name: fuga
  id: 2
  revision: a047baca-83d3-4d7f-8b10-c39abc7b1b45
  archive type: db
  project created at: 2019-03-27 17:20:17 +0900
  revision updated at: 2019-03-27 17:20:17 +0900
```

`digdag projects fuga`


```
  name: fuga
  id: 2
  revision: a047baca-83d3-4d7f-8b10-c39abc7b1b45
  archive type: db
  project created at: 2019-03-27 17:20:17 +0900
  revision updated at: 2019-03-27 17:20:17 +0900
```

**Comment by yoyama** [comment]
@hiroyuki-sato Sorry for my broken reply.
Yes that's right. I think keeping consistency with other commands as much as possible is better for users.

**Comment by hiroyuki-sato** [comment]
@yoyama I simply implemented that feature.  Please take a look when you have time.

I'm wondering whether it should output like "no much" if proj_name doesn't match any project.

**Comment by yoyama** [comment]
Can you use DigdagClient.getProject(String name) if the name argument is specified?

**Comment by hiroyuki-sato** [comment]
@yoyama I reimplemented using `DigdagClient.getProject(String name)`. Please take a look when you get a chance.

File: digdag-cli/src/main/java/io/digdag/cli/Main.java [hunk 3]
```diff
@@ -133,6 +134,7 @@ protected void configure()
         jc.addCommand("upload", injector.getInstance(Upload.class));
         jc.addCommand("download", injector.getInstance(Download.class));
 
+        jc.addCommand("projects", injector.getInstance(ShowProjects.class), "projects");
```
Comment (yoyama) [inline digdag-cli/src/main/java/io/digdag/cli/Main.java]: Third argument is alias. You can set alias same as workflow -> workflows or as secrets -> secret.

File: digdag-cli/src/main/java/io/digdag/cli/client/ShowProjects.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        ln("  revision: %s", project.getRevision());
+        ln("  archive type: %s", project.getArchiveType());
+        ln("  project created at: %s", TimeUtil.formatTime(project.getCreatedAt()));
+        ln("  revision updated at: %s", TimeUtil.formatTime(project.getUpdatedAt()));
+        ln("");
+    }
+
+    public SystemExitException usage(String error)
+    {
+        err.println("Usage: " + programName + " projects [project]");
```
Comment (yoyama) [inline digdag-cli/src/main/java/io/digdag/cli/client/ShowProjects.java]: [project] -> [name] may be better as workflows command.

File: digdag-cli/src/main/java/io/digdag/cli/Main.java [hunk 5]
```diff
@@ -313,6 +315,7 @@ private SystemExitException usage(String error)
         err.println("    reschedule <schedule-id>           skip sessions of a schedule to a future time");
         err.println("    reschedule <project-name> <name>   skip sessions of a schedule to a future time");
         err.println("    log <attempt-id>                   show logs of a session attempt");
+        err.println("    projects                           show projects");
```
Comment (yoyama) [inline digdag-cli/src/main/java/io/digdag/cli/Main.java]: Add [name] may be better?

File: digdag-tests/src/test/java/acceptance/CliShowIT.java [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+import java.nio.file.Path;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.containsString;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
+import static utils.TestUtils.copyResource;
+import static utils.TestUtils.main;
+
+public class CliShowIT
```
Comment (yoyama) [inline digdag-tests/src/test/java/acceptance/CliShowIT.java]: On the class name CliShowProjectsIT may be better?

**Comment by yoyama** [comment]
@hiroyuki-sato Thanks for your works. I think it is basically good. But I had a few comments.
Could you check them if possible?

**Comment by hiroyuki-sato** [comment]
@yoyama Fixed your comments. PTAL again.

File: digdag-tests/src/test/java/acceptance/CliShowProjectsIT.java [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                    "-c", config.toString(),
+                    "-e", server.endpoint());
+            assertThat(pushStatus.errUtf8(), pushStatus.code(), is(0));
+        }
+
+        CommandStatus showProjectStatus = main("projects","not_exist_project",
+                "-c", config.toString(),
+                "-e", server.endpoint());
+
+        assertThat(showProjectStatus.errUtf8(), showProjectStatus.code(), is(0));
```
Comment (yoyama) [inline digdag-tests/src/test/java/acceptance/CliShowProjectsIT.java]: Is exit code 0 is right when resource not found?
It looks like other commands such as 'workflows' return non-zero if resource not found. 

```
$ digdag workflows not_exist_workflow || echo fail
2019-04-10 14:35:53 +0900: Digdag v0.9.35
error: Resource not found: project not found: not_exist_workflow
fail
```

**Comment by yoyama** [comment]
@hiroyuki-sato Thank you for your modification. I made one more comment. Could you check it again?

**Comment by hiroyuki-sato** [comment]
@yoyama Thanks!, Fixed. Please take a look again.


```
./pkg/digdag-0.9.36-SNAPSHOT.jar projects fuga
2019-04-10 18:51:30 +0900: Digdag v0.9.36-SNAPSHOT
error: Project 'fuga' does not exist.
```

And `exit status 1`

**Comment by hiroyuki-sato** [comment]
Thank you!

**Pull request merged**